File: 2/main.py
def main():
    with open("input.txt") as f:
        lines = f.readlines()

    commands = [(cmd, int(x)) for cmd, x in (line.split() for line in lines)]
    h, d, a = 0, 0, 0

    for cmd, x in commands:
        if cmd == "forward":
            h += x
            d += a * x
        elif cmd == "down":
            a += x
        elif cmd == "up":
            a -= x

    print(h, "*", d, "=", h * d)


if __name__ == "__main__":
    main()

File: 2/main.rs
use std::fs::File;
use std::io::{self, BufRead};

fn main() {
    let file = File::open("input.txt").unwrap();
    let lines: Vec<String> = io::BufReader::new(file)
        .lines()
        .map(|l| l.unwrap())
        .collect();

    let (mut h, mut d, mut a) = (0, 0, 0);

    for line in lines {
        let mut parts = line.split(" ");
        let cmd = parts.next().unwrap();
        let x: i32 = parts.next().unwrap().parse().unwrap();
        match cmd {
            "forward" => {
                h += x;
                d += a * x;
            }
            "down" => a += x,
            "up" => a -= x,
            _ => unreachable!(),
        };
    }

    println!("{} * {} = {}", h, d, h * d);
}
